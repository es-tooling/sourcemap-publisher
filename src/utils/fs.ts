import {fdir} from 'fdir';
import path from 'node:path';
import {mkdir, rm, stat, cp} from 'node:fs/promises';

export async function getSourceFilesFromPaths(
  cwd: string,
  paths: string[]
): Promise<string[]> {
  const crawler = new fdir();
  const files = await crawler
    .withFullPaths()
    .exclude((_dirName, dirPath) => {
      return !paths.some((p) => dirPath.startsWith(p));
    })
    .filter((file) => {
      return (
        paths.some((p) => file.startsWith(p)) &&
        (file.endsWith('.js') ||
          (!file.endsWith('.d.ts') && file.endsWith('.ts')))
      );
    })
    .crawl(cwd)
    .withPromise();
  return files;
}

export async function getTempDir(cwd: string, name: string): Promise<string> {
  const tempDir = path.join(cwd, name);

  await rm(tempDir, {force: true, recursive: true});
  await mkdir(tempDir);

  return tempDir;
}

export async function copyRelativeFilesToDir(
  files: string[],
  sourceDir: string,
  targetDir: string
): Promise<void> {
  for (const file of files) {
    const sourcePath = path.join(sourceDir, file);
    const targetPath = path.join(targetDir, file);
    try {
      await stat(sourcePath);
      await cp(sourcePath, targetPath, {recursive: true});
    } catch {
      continue;
    }
  }
}
