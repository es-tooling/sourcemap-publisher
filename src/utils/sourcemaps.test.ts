import {suite, test, expect, beforeEach, afterEach, vi} from 'vitest';
import {
  createExternalSourcemapUrl,
  ExtractedSourceMapSuccess,
  extractSourceMap,
  updateSourceMapUrls
} from './sourcemaps.js';
import type {PackageJson} from './package-json.js';
import {mkdtemp, readFile, rm, writeFile} from 'fs/promises';
import path from 'path';
import {tmpdir} from 'os';

suite('createExternalSourcemapUrl', () => {
  test('should template url', () => {
    const file = 'foo/bar.js.map';
    const pkg: PackageJson = {
      name: 'test-package',
      version: '1.0.0-sourcemaps',
      files: []
    };
    expect(createExternalSourcemapUrl(file, pkg)).toBe(
      'https://unpkg.com/test-package@1.0.0-sourcemaps/foo/bar.js.map'
    );
  });
});

const writeFiles = async (
  files: Record<string, string>,
  cwd: string
): Promise<void> => {
  for (const [name, content] of Object.entries(files)) {
    await writeFile(path.join(cwd, name), content);
  }
};

suite('updateSourceMapUrls', () => {
  let pkg: PackageJson;
  let tempDir: string;
  let files: Record<string, string>;

  beforeEach(async () => {
    pkg = {
      name: 'test-package',
      version: '1.0.0',
      files: []
    };

    tempDir = await mkdtemp(path.join(tmpdir(), 'smpub'));

    files = {
      'foo.js': `
// This is a test file
//# sourceMappingURL=foo.js.map`,
      'bar.js': `
// This is a test file
//# sourceMappingURL=bar.js.map`,
      'bar.js.map': '// x',
      'foo.js.map': '// x'
    };

    await writeFiles(files, tempDir);
  });

  afterEach(async () => {
    await rm(tempDir, {recursive: true, force: true});
    vi.restoreAllMocks();
  });

  test('replaces urls with CDN urls', async () => {
    const sourceMaps: ExtractedSourceMapSuccess[] = [];

    for (const [file, source] of Object.entries(files)) {
      if (file.endsWith('.map')) {
        continue;
      }
      const substr = 'sourceMappingURL=';
      const rangeStart = source.indexOf(substr) + substr.length;
      sourceMaps.push({
        success: true,
        path: path.join(tempDir, `${file}.map`),
        source: path.join(tempDir, file),
        range: [rangeStart, source.length]
      });
    }

    await updateSourceMapUrls(tempDir, sourceMaps, pkg);

    for (const sourceMap of sourceMaps) {
      const contents = await readFile(sourceMap.source, 'utf8');

      expect(contents).toMatchSnapshot();
    }
  });
});

suite('extractSourceMap', () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = await mkdtemp(path.join(tmpdir(), 'smpub'));
  });

  afterEach(async () => {
    await rm(tempDir, {recursive: true, force: true});
  });

  test('errors when file does not exist', async () => {
    const source = path.join(tempDir, 'non-existent.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'could not load source file'
    });
  });

  test('errors when no sourcemap URL', async () => {
    await writeFiles(
      {
        'foo.js': '// foo'
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'no sourcemap found'
    });
  });

  test('errors when absolute sourcemap URL', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
//# sourceMappingURL=/absolute/sourcemap.js.map`
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'absolute and external URLs not supported'
    });
  });

  test('errors when external sourcemap URL', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
//# sourceMappingURL=https://example.com/sourcemap.js.map`
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'absolute and external URLs not supported'
    });
  });

  test('errors when data URL', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
//# sourceMappingURL=data:application/json;base64,woowoo`
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'data URLs not supported'
    });
  });

  test('errors when sourcemap does not exist', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
//# sourceMappingURL=foo.js.map`
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'sourcemap not found'
    });
  });

  test('ignores sourcemaps in weird places', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
303;
//# sourceMappingURL=funky-sourcemaps.js.map
808;`
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = await extractSourceMap(source);

    expect(result).toEqual({
      source,
      success: false,
      reason: 'no sourcemap found'
    });
  });

  test('retrieves sourcemap URL', async () => {
    const contents = `// foo
//# sourceMappingURL=foo.js.map`;
    await writeFiles(
      {
        'foo.js': contents,
        'foo.js.map': '// foo'
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = (await extractSourceMap(
      source
    )) as ExtractedSourceMapSuccess;

    result.source = result.source.replace(tempDir, 'TEMP_DIR');
    result.path = result.path.replace(tempDir, 'TEMP_DIR');
    expect(result).toMatchSnapshot();
    expect(contents.slice(result.range[0], result.range[1])).toBe('foo.js.map');
  });
});

suite('extractSourceMaps', () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = await mkdtemp(path.join(tmpdir(), 'smpub'));
  });

  afterEach(async () => {
    await rm(tempDir, {recursive: true, force: true});
  });

  test('extracts sourcemaps from files', async () => {
    await writeFiles(
      {
        'foo.js': `// foo
//# sourceMappingURL=foo.js.map`,
        'foo.js.map': '// foo',
        'bar.js': '// i have no sourcemap'
      },
      tempDir
    );
    const source = path.join(tempDir, 'foo.js');
    const result = (await extractSourceMap(
      source
    )) as ExtractedSourceMapSuccess;

    result.source = result.source.replace(tempDir, 'TEMP_DIR');
    result.path = result.path.replace(tempDir, 'TEMP_DIR');
    expect(result).toMatchSnapshot();
  });
});
