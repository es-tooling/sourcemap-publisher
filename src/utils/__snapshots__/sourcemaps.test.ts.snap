// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`updateSourceMapUrls > replaces urls with CDN urls 1`] = `
"
// This is a test file
//# sourceMappingURL=https://unpkg.com/test-package@1.0.0/foo.js.map"
`;

exports[`updateSourceMapUrls > replaces urls with CDN urls 2`] = `
"
// This is a test file
//# sourceMappingURL=https://unpkg.com/test-package@1.0.0/foo.js.map"
`;

exports[`updateSourceMapUrls > replaces urls with CDN urls 3`] = `"// x"`;

exports[`updateSourceMapUrls > replaces urls with CDN urls 4`] = `"// x"`;

exports[`updateSourceMapUrls > skips on non-existent sourcemap 1`] = `
"
// This is a test file
//# sourceMappingURL=https://unpkg.com/test-package@1.0.0/foo.js.map"
`;
